### DISCUSSION TITLE
How this repo makes Newtonsoft works on AOT?

### DISCUSSION THREAD

**Discussion by gustavopsantos** [applejag/Newtonsoft.Json-for-Unity#147] [Q&A] [answered]
If I try to use normal Newtonsoft on Unity + AOT Scripting Backend, its not secret, I'll get that exception: `PlatformNotSupportedException: Operation is not supported on this platform. (referring to NewtonSoft's use of System.Reflection.Emit.DynamicMethod..ctor)`

So how this repo handles this?
It is normal newtonsoft trying to Emit IL code right?
But how about Newtonsoft.Json-for-Unity? It relies on reflection? How does this works under the hood? Maybe you can point me the core classes that makes magic happen.
Thanks in advance!

**Comment by applejag** [comment, marked as answer]
Excellent question!

On a higher level, I've filtered out the code that uses System.Reflection.Emit and other JIT-based code via a new build target. One key example of where this filtering is applied is at:

https://github.com/jilleJr/Newtonsoft.Json-for-Unity/blob/2a26db68c98003f28749203378a3863afa5e6885/Src/Newtonsoft.Json/Serialization/DefaultContractResolver.cs#L1405-L1426

And:

https://github.com/jilleJr/Newtonsoft.Json-for-Unity/blob/2a26db68c98003f28749203378a3863afa5e6885/Src/Newtonsoft.Json/Serialization/JsonTypeReflector.cs#L514-L531

The `UNITY_LTS` pragma flag has been sprinkled across the repo to filter out some features. Newtonsoft.Json is built in such a way that it supports a lot of targets by filtering via define constants (`#if` statements) for a lot of things, so I could pick and choose what's supported in Unity, and then do some minor adjustments here and there for the remaining things that I couldn't filter out appropriately via define constants:

https://github.com/jilleJr/Newtonsoft.Json-for-Unity/blob/2a26db68c98003f28749203378a3863afa5e6885/Src/Newtonsoft.Json/Newtonsoft.Json.csproj#L69-L75

The official Newtonsoft.Json barely emits any IL explicitly, but instead uses some other constructs where you can convert syntax trees (AST) into IL quite easily, so it can on-the-fly generate functions that reads/writes to an object's fields which is increadibly faster than reading/writing via reflection.

This means that the AOT build I've set up for Unity will never be as fast as Newtonsoft.Json running bare metal on a real CLR, but it's still increadibly fast. It only performs reflection to find the fields and properties once per type and caches the field's pointers in contracts.

Originally the changes were quite minimal and were based on a comment by @dngulin (https://github.com/JamesNK/Newtonsoft.Json/issues/1440#issuecomment-430510160), and can be found here: https://github.com/JamesNK/Newtonsoft.Json/pull/2044/files

However as time went on, I found some more issues with IL2CPP and fixed them outside of that PR, as well as added other features such as:

- AotHelper from SaladLab (https://github.com/SaladLab/Json.Net.Unity3D)
- Unit testing inside Unity via Unity's test runner inside a CI pipeline (running inside CircleCI https://app.circleci.com/pipelines/github/jilleJr/Newtonsoft.Json-for-Unity)
  - in the Unity editor
  - in playmode on both Mono and IL2CPP builds

A full list of all changes can be found here: https://github.com/JamesNK/Newtonsoft.Json/compare/JamesNK:13.0.1...jilleJr:2a26db6
The full compare is a bit tedious to use due to the added Unity tests and CI stuff. Just try to scroll down to the `Src/Newtonsoft.Json/*` part, which is almost at the bottom.

**Reply by gustavopsantos** [reply]
Thnaks for blazing fast reply <3
